src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -359,6 +383,217 @@ def to_Spec {f : A} (m : ℕ) (f_deg : f ∈ 𝒜 m) :
+suffices α_sq_mem : α * α ∈ q.1, by { cases q.2.mem_or_mem α_sq_mem; assumption },
+have α_sq_eq' : (α * α : away f) = ∑ (j : ℕ) in range (2 * m + 1),
+  mk ((proj 𝒜 i) a ^ j * (proj 𝒜 i) b ^ (2 * m - j) * ↑((2 * m).choose j)) ⟨f ^ (2 * i), ⟨_, rfl⟩⟩,
+begin
+  rw [←localization.mk_sum, ←add_pow, ←map_add],
+  simpa only [degree_zero_part.coe_mul, subtype.coe_mk, mk_mul, two_mul, pow_add],
+end,
+have α_sq_eq : α * α = ∑ j in (range (2 * m + 1)).attach,
+  ⟨mk ((proj 𝒜 i) a ^ j.1 * (proj 𝒜 i) b ^ (2 * m - j.1) * ↑((2 * m).choose j.1))
+    ⟨f ^ (2 * i), ⟨_, rfl⟩⟩,
+    ⟨2 * i, ⟨⟨_, begin
+      rw show m * (2 * i) = ((j.1*i) + (2*m-j.1)*i + 0),
+      { zify [show j.1 ≤ 2 * m, by linarith [finset.mem_range.mp j.2]], linarith, },
+      mem_tac,
+    end⟩, rfl⟩⟩⟩,
+begin
+  rw [subtype.ext_iff, degree_zero_part.coe_sum, degree_zero_part.coe_mul],
+  rwa ←finset.sum_attach at α_sq_eq',
+end,
+α_sq_eq.symm ▸ ideal.sum_mem _
+begin
+  rintros ⟨k, hk⟩ _,
+  rw finset.mem_range at hk,
+  have hk' : k ≤ 2 * m := by linarith,
+
+  by_cases ineq : m ≤ k,
+  { set α := (⟨mk ((proj 𝒜 i) a ^ m) ⟨f^i, ⟨i, rfl⟩⟩, ⟨i, ⟨_, by mem_tac⟩, rfl⟩⟩ : A⁰_ f_deg),
+    set β := (⟨mk ((proj 𝒜 i) a ^ (k - m) *
+        (proj 𝒜 i) b ^ (2 * m - k) * (2*m).choose k) ⟨f^i, ⟨i, rfl⟩⟩, begin
+          refine ⟨i, ⟨_, _⟩, rfl⟩,
+          rw show m * i = ((k - m) * i) + ((2*m-k) * i) + 0,
+          { zify [hk, hk'], linarith, },
+          mem_tac,
+        end⟩ : A⁰_ f_deg),
+    suffices : α * β ∈ q.1,
+    { convert this,
+      simp only [mk_mul, two_mul, pow_add],
+      congr' 1,
+      rw [←mul_assoc, ←mul_assoc, ←pow_add],
+      congr' 3,
+      zify [hk, hk'],
+      linarith, },
+    exact ideal.mul_mem_right _ _ (ha _), },
+    { have ineq' : k ≤ m := by linarith only [ineq],
+      set α := (⟨mk ((proj 𝒜 i) b ^ m) ⟨f^i, ⟨_, rfl⟩⟩, ⟨i, ⟨_, by mem_tac⟩, rfl⟩⟩ : A⁰_ f_deg),
+      set β := (⟨mk ((proj 𝒜 i) a ^ k * (proj 𝒜 i) b ^ (m - k) * ((2 * m).choose k))
+        ⟨f^i, ⟨_, rfl⟩⟩, begin
+          refine ⟨_, ⟨_, _⟩, rfl⟩,
+          rw ← show k * i + (m - k) * i + 0 = m * i,
+          { zify [hk, hk', ineq'], linarith, },
+          mem_tac,
+        end⟩ : A⁰_ f_deg),
+      suffices : α * β ∈ q.1,
+      { convert this,
+        simp only [mk_mul, two_mul, pow_add],
+        congr' 1,
+        rw [show ∀ (a b c d : A), a * (b * c * d) = b * (a * c) * d, by {intros, ring}, ←pow_add,
+          ←nat.add_sub_assoc ineq', ←two_mul], },
+      exact ideal.mul_mem_right _ _ (hb _), },
+end
+
+lemma carrier.smul_mem (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) (c x : A) (hx : x ∈ carrier q) :
+  c • x ∈ carrier q :=
+begin
+  classical,
+  revert c,
+  refine direct_sum.decomposition.induction_on 𝒜 _ _ _,
+  { rw zero_smul, exact carrier.zero_mem hm _, },
+  { rintros n ⟨a, ha⟩ i,
+    by_cases ineq1 : n ≤ i,
+    { have eq1' : proj 𝒜 i (a * x) = a * (proj 𝒜 (i - n) x),
+      { have := @direct_sum.coe_decompose_mul_add_of_left_mem _ _ _ _ _ _ _ _ 𝒜 _ a x _ (i - n) ha,
```

Comment:
I think it's worth extracting a lemma analogous to [direct_sum.coe_decompose_mul_add_of_left_mem](https://leanprover-community.github.io/mathlib_docs/ring_theory/graded_algebra/basic.html#direct_sum.coe_decompose_mul_add_of_left_mem) but says
```
direct_sum.decompose 𝒜 (a * b) j = if i ≤ j then a * direct_sum.decompose 𝒜 b (j - i) else 0
```
when `ι` is `ℕ`.

Reply:
This lemma has been created and also in a separate pr #16083